File: backend/providers/ollama_provider.py
import json
from typing import AsyncIterator, Optional

import httpx

from .base import LLMProvider


class OllamaProvider(LLMProvider):
    """
    Local-only provider. Requires Ollama running (`ollama serve`) with a
    vision-capable model pulled, e.g. `ollama pull llava`.
    """

    name = "ollama"

    def __init__(self, model: str = "llava", base_url: str = "http://127.0.0.1:11434"):
        self.model = model
        self.base_url = base_url

    async def stream_vision_response(
        self,
        image_data_url: str,
        transcript_context: str,
        system_prompt: str,
        api_key: Optional[str] = None,
    ) -> AsyncIterator[str]:
        _, b64data = image_data_url.split(",", 1)

        user_text = (
            f"{system_prompt}\n\nRecent transcript:\n{transcript_context}\n\n"
            "Respond with the most useful, concise assistance for what's happening right now."
        )

        payload = {
            "model": self.model,
            "prompt": user_text,
            "images": [b64data],
            "stream": True,
        }

        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream("POST", f"{self.base_url}/api/generate", json=payload) as resp:
                async for line in resp.aiter_lines():
                    if not line:
                        continue
                    data = json.loads(line)
                    if data.get("response"):
                        yield data["response"]
                    if data.get("done"):
                        break

File: android/app/src/main/java/com/wisp/app/providers/OllamaProvider.kt
package com.wisp.app.providers

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException

/**
 * Unlike every other provider here, points at a server the user runs
 * themselves rather than a cloud API -- and unlike the desktop app (where
 * Ollama is typically localhost), a phone has no "local" Ollama, so
 * baseUrl must be a reachable address on the same network (see
 * WispSettings.ollamaBaseUrl's default and doc comment).
 *
 * Ollama's /api/generate streams newline-delimited JSON objects, NOT
 * text/event-stream ("data: " lines) like the other 4 providers, so this
 * can't reuse streamSse() -- same distinction as the desktop app's
 * ollama_provider.py vs. its SSE-based siblings.
 */
class OllamaProvider(private val baseUrl: String, private val model: String = "llava") : LLMProvider {
    override fun streamVisionResponse(
        imageBase64: String,
        mimeType: String,
        transcriptContext: String,
        systemPrompt: String,
        apiKey: String,
    ): Flow<String> = flow {
        val prompt = "$systemPrompt\n\n${buildUserText(transcriptContext)}"
        val body = JSONObject()
            .put("model", model)
            .put("prompt", prompt)
            .put("images", JSONArray().put(imageBase64))
            .put("stream", true)

        val request = Request.Builder()
            .url("$baseUrl/api/generate")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = sharedHttpClient.newCall(request).execute()
        response.use { resp ->
            if (!resp.isSuccessful) {
                throw IOException("HTTP ${resp.code}: ${resp.body?.string().orEmpty()}")
            }
            val source = resp.body?.source() ?: return@use
            while (!source.exhausted()) {
                val line = source.readUtf8Line() ?: break
                if (line.isBlank()) continue
                val json = JSONObject(line)
                json.optString("response").takeIf { it.isNotEmpty() }?.let { emit(it) }
                if (json.optBoolean("done", false)) break
            }
        }
    }
}
